cmd/file.d_e2e_new_test.go

```diff
@@ -0,0 +1,253 @@
+	matches := test.GetMatches(t, logFilePattern)
+	assert.True(t, len(matches) > 0, "There are no files")
+	require.Equal(t, f.count*f.lines, test.CountLines(t, logFilePattern))
+}
+
+// Struct for http-file plugin e2e test
+type httpFile struct {
+	filesDir          string
+	outputNamePattern string
+	count             int
+	lines             int
+	retTime           string
+}
+
+func (h *httpFile) AddConfigSettings(t *testing.T, conf *cfg.Config, pipeName string) {
+	h.filesDir = t.TempDir()
+	output := conf.Pipelines[pipeName].Raw.Get("output")
+	output.Set("target_file", h.filesDir+h.outputNamePattern)
+	output.Set("retention_interval", h.retTime)
+}
+
+func (h *httpFile) Send(t *testing.T) {
+	wg := &sync.WaitGroup{}
+	wg.Add(h.count)
+	for i := 0; i < h.count; i++ {
+		go func() {
+			cl := http2.DefaultClient
+			for i := 0; i < h.lines; i++ {
+				rd := bytes.NewReader([]byte(`{"first_field":"second_field"}`))
+				req, err := http2.NewRequest(http2.MethodPost, "http://localhost:9200/", rd)
+				assert.Nil(t, err, "bad format http request")
+				_, err = cl.Do(req)
+				if err != nil {
+					fmt.Println(err)
+				}
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	// waiting for files to be processed
+	time.Sleep(time.Second * 10)
+}
+
+func (h *httpFile) Validate(t *testing.T) {
+	logFilePattern := fmt.Sprintf("%s/%s*%s", h.filesDir, "file-d", ".log")
+	matches := test.GetMatches(t, logFilePattern)
+	assert.True(t, len(matches) > 0, "There are no files")
+	require.Equal(t, h.count*h.lines, test.CountLines(t, logFilePattern))
+}
+
+// Struct for kafka-file plugin e2e test
+type kafkaFile struct {
+	topic             string
+	broker            string
+	filesDir          string
+	outputNamePattern string
+	lines             int
+	retTime           string
+	partition         int
+}
+
+func (k *kafkaFile) AddConfigSettings(t *testing.T, conf *cfg.Config, pipeName string) {
+	k.filesDir = t.TempDir()
+	output := conf.Pipelines[pipeName].Raw.Get("output")
+	output.Set("target_file", k.filesDir+k.outputNamePattern)
+	output.Set("retention_interval", k.retTime)
+}
+
+func (k *kafkaFile) Send(t *testing.T) {
+	time.Sleep(10 * time.Second)
+	connections := make([]*kafka.Conn, k.partition)
+	for i := 0; i < k.partition; i++ {
+		conn, err := kafka.DialLeader(context.Background(), "tcp", "localhost:9092", k.topic, i)
+		if err != nil {
+			log.Fatal("failed to dial leader:", err)
+		}
+		conn.SetWriteDeadline(time.Now().Add(20 * time.Second))
+		connections[i] = conn
+	}
+
+	for i := 0; i < k.lines; i++ {
+		for _, conn := range connections {
+			_, err := conn.WriteMessages(
+				kafka.Message{Value: []byte(`{"key":"value"}`)},
+			)
+			if err != nil {
+				log.Fatal("failed to write messages:", err)
+			}
+		}
+	}
+
+	for _, conn := range connections {
+		if err := conn.Close(); err != nil {
+			log.Fatal("failed to close writer:", err)
+		}
+	}
+	// waiting for files to be processed
+	time.Sleep(time.Second * 10)
```

Comment:
let's try not to use sleep in tests

Reply:
we don't have finer control over event processing time

Reply:
you can check number of events which is processed by the pipeline and wait until it gets to the count you expecting